File: service/src/main/java/bridge/socket/ServerSocketHandler.java
package bridge.socket;

import bridge.BridgeService;
import io.netty.channel.Channel;
import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.SimpleChannelInboundHandler;
import io.netty.handler.codec.http.websocketx.TextWebSocketFrame;
import io.netty.handler.codec.http.websocketx.WebSocketFrame;
import io.netty.util.AttributeKey;
import packages.BridgePackage;
import packages.GreetingPackage;
import util.UtilNetty;

import java.net.InetSocketAddress;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;

public class ServerSocketHandler extends SimpleChannelInboundHandler<WebSocketFrame> {

	private static final AttributeKey<String> serverInfoKey = AttributeKey.newInstance("serverinfo");

	public static final Map<String, Channel> connectedChannels = new ConcurrentHashMap<>();

	public static void broadcast(BridgePackage pckg) {
		connectedChannels.values().forEach(channel -> send(channel, UtilNetty.toFrame(pckg)));
	}

	public static void send(Channel channel, BridgePackage pckg) {
		send(channel, UtilNetty.toFrame(pckg));
	}

	public static void send(Channel channel, TextWebSocketFrame frame) {
		channel.writeAndFlush(frame, channel.voidPromise());
	}

	@Override
	protected void channelRead0(ChannelHandlerContext ctx, WebSocketFrame msg) {
		if (msg instanceof TextWebSocketFrame) {
			BridgePackage thePitPackage = UtilNetty.readFrame((TextWebSocketFrame) msg);
			Channel channel = ctx.channel();
			if (thePitPackage instanceof GreetingPackage) {
				if (channel.hasAttr(serverInfoKey)) {
					System.out.println("Some channel tries to authorize, but it already in system!");
					return;
				}
				GreetingPackage pckg = (GreetingPackage) thePitPackage;
				if (connectedChannels.containsKey(pckg.getServerName())) {
					System.out.println("Channel want to register as " + pckg.getServerName() + ", but this name already in use!");
					ctx.close();
					return;
				}
				if (!pckg.getPassword().equals(BridgeService.PASSWORD)) {
					System.out.println("Channel provided bad password: " + pckg.getPassword());
					if (channel.remoteAddress() instanceof InetSocketAddress) {
						System.out.println(channel.remoteAddress().toString());
					}
					ctx.close();
					return;
				}
				channel.attr(serverInfoKey).set(pckg.getServerName());
				connectedChannels.put(pckg.getServerName(), channel);
			} else {
				if (!channel.hasAttr(serverInfoKey)) {
					System.out.println("Some channel tries to send packet without authorization!");
					if (channel.remoteAddress() instanceof InetSocketAddress) {
						System.out.println(channel.remoteAddress().toString());
					}
					ctx.close();
					return;
				}
				String info = channel.attr(serverInfoKey).get();
				Optional.ofNullable(BridgeService.HANDLER_MAP.get(thePitPackage.getClass()))
						.ifPresent(handler -> handler.handle(channel, info, thePitPackage));
			}
		}
	}

	@Override
	public void channelInactive(ChannelHandlerContext ctx) {
		Channel channel = ctx.channel();
		if (channel.hasAttr(serverInfoKey)) {
			String name = channel.attr(serverInfoKey).get();
			connectedChannels.remove(name);
			System.out.println("Server disconnected! " + name);
		}
	}

}

File: service/src/main/kotlin/socket/ServerSocketHandler.kt
package socket

import handlerMap
import password
import io.netty.channel.Channel
import io.netty.channel.ChannelHandlerContext
import io.netty.channel.SimpleChannelInboundHandler
import io.netty.handler.codec.http.websocketx.TextWebSocketFrame
import io.netty.handler.codec.http.websocketx.WebSocketFrame
import io.netty.util.AttributeKey
import me.reidj.common.`package`.BridgePackage
import me.reidj.common.`package`.GreetingPackage
import me.reidj.common.util.UtilNetty
import java.net.InetSocketAddress
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.function.Consumer

object ServerSocketHandler : SimpleChannelInboundHandler<WebSocketFrame>() {

    private val serverInfoKey = AttributeKey.newInstance<String>("serverinfo")

    private val connectedChannels: MutableMap<String, Channel> = ConcurrentHashMap()

    fun broadcast(pckg: BridgePackage) {
        connectedChannels.values.forEach(Consumer { channel: Channel ->
            send(channel, UtilNetty.toFrame(pckg))
        })
    }

    fun send(channel: Channel, pckg: BridgePackage) {
        send(channel, UtilNetty.toFrame(pckg))
    }

    fun send(channel: Channel, frame: TextWebSocketFrame?) {
        channel.writeAndFlush(frame, channel.voidPromise())
    }

    override fun channelRead0(ctx: ChannelHandlerContext, msg: WebSocketFrame?) {
        if (msg is TextWebSocketFrame) {
            val thePitPackage: BridgePackage = msg.let { UtilNetty.readFrame(it) }
            val channel = ctx.channel()
            if (thePitPackage is GreetingPackage) {
                if (channel.hasAttr(serverInfoKey)) {
                    println("Some channel tries to authorize, but it already in system!")
                    return
                }
                val pckg: GreetingPackage = thePitPackage
                if (connectedChannels.containsKey(pckg.serverName)) {
                    println("Channel want to register as " + pckg.serverName + ", but this name already in use!")
                    ctx.close()
                    return
                }
                if (pckg.password != password) {
                    println("Channel provided bad password: $password")
                    if (channel.remoteAddress() is InetSocketAddress) {
                        println(channel.remoteAddress().toString())
                    }
                    ctx.close()
                    return
                }
                channel.attr(serverInfoKey).set(pckg.serverName)
                connectedChannels[pckg.serverName] = channel
            } else {
                if (!channel.hasAttr(serverInfoKey)) {
                    println("Some channel tries to send packet without authorization!")
                    if (channel.remoteAddress() is InetSocketAddress) {
                        println(channel.remoteAddress().toString())
                    }
                    ctx.close()
                    return
                }
                val info = channel.attr(serverInfoKey).get()
                Optional.ofNullable(handlerMap[thePitPackage::class.java])
                    .ifPresent { it.handle(channel, info, thePitPackage as Nothing) }
            }
        }
    }

    override fun channelInactive(ctx: ChannelHandlerContext) {
        val channel = ctx.channel()
        if (channel.hasAttr(serverInfoKey)) {
            val name = channel.attr(serverInfoKey).get()
            connectedChannels.remove(name)
            println("Server disconnected! $name")
        }
    }
}